src/Native/System.Native/pal_networking.cpp

```diff
@@ -379,7 +380,119 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR && (!HAVE_GETHOSTBYNAME_R && !HAVE_GETHOSTBYADDR_R)
+#define BUFFER_OVERFLOW 255
+static int copy_hostent(struct hostent* from, struct hostent* to,
+                        char* buffer, size_t buflen, int* error)
+{
+    int i, len;
+    char* ptr = buffer;
+
+    *error = 0;
+    to->h_addrtype = from->h_addrtype;
+    to->h_length = from->h_length;
+    to->h_name = static_cast<char*>(ptr);
+
+    /* copy hostname to buffer */
+    len = static_cast<int>(strlen(from->h_name)+1);
+    SafeStringCopy(ptr, len, from->h_name);
+    ptr += len;
+
+    /* copy aliases to buffer */
+    to->h_aliases = reinterpret_cast<char**>(ptr);
+    for(i = 0; from->h_aliases[i]; i++);
+    ptr += (i+1) * static_cast<int>(sizeof(char*));
+
+    for(i = 0; from->h_aliases[i]; i++)
+    {
+        len = static_cast<int>(strlen(from->h_aliases[i])+1);
+        if (static_cast<size_t>((ptr-buffer)+len) < buflen)
+        {
+            to->h_aliases[i] = ptr;
+            SafeStringCopy(ptr, len, from->h_aliases[i]);
```

Comment:
As noted earlier, the second arg is supposed to be the destination size, so it needs to be the computed amount of space remaining in the buffer.